deprecate opensea security alerts check and settings

segun opened the pull request:
## Explanation

### Background
The new Blockaid integration should replace the existing OpenSea + Blockaid security provider feature. 

### Acceptance Criteria
1. Users that have the Opensea + Blockaid security provider enabled should have the new Blockaid feature enabled by default when it's rolled out
2. Users that have the Opensea + Blockaid security provider disabled should have the new Blockaid feature disabled by default when it's rolled out
3. The Opensea + Blockaid security provider should be disabled for all users - No data should be sent anymore to the API and no warning should be shown as a result of this feature
4. The Opensea + Blockaid security provider should not appear within settings anymore 
5. Metrics defined here https://github.com/MetaMask/MetaMask-planning/issues/527 should be triggered when users have the Blockaid feature enabled by default. (`blockaid` value should be added to the `security_providers` user property when it's enabled to users and Settings updated event should be triggered)
6. The `opensea` value should be removed from `security_providers` user property when it's disabled/deprecated for that user

### Out of scope 
We'll communicate users of this migration as part of the what's new text.

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/926

## Screenshots/Screencaps

### Before

**UI ScreenShot 1**
<img width="930" alt="Screenshot 2023-07-14 at 15 14 01" src="https://github.com/MetaMask/metamask-extension/assets/44811/38e49818-8255-4bd0-bd4f-65eca5c808c2">

**Transaction Data 1**
<img width="1256" alt="Screenshot 2023-07-28 at 11 40 51" src="https://github.com/MetaMask/metamask-extension/assets/44811/d0fea47f-5eed-41b9-8c17-d820b4083afb">

### After
**UI ScreenShot 2**
<img width="944" alt="Screenshot 2023-07-14 at 15 16 59" src="https://github.com/MetaMask/metamask-extension/assets/44811/587f01c9-76ee-42fb-865e-4be1ba9bf143">

**Transaction Data 2**
<img width="1262" alt="Screenshot 2023-07-28 at 11 43 17" src="https://github.com/MetaMask/metamask-extension/assets/44811/5d1453e3-7bcb-45bd-b64a-225baffd2931">

## Manual Testing Steps

1. Start MM using `yarn start`
2. Open Settings
3. Tab Experimental
4. You should see the screenshot like in **UI ScreenShot 1**
5. Send a transaction from test-dapp
6. Open console and type txData.securityProviderResponse, it should be null as in screenshot **Transaction Data 1**
7. Stop MM and restart using `yarn start:flask`
8. Open Settings
9. Tab Experimental
10. You should see the screenshot like in **UI ScreenShot 2**
11. Send a transaction from test-dapp
12. Open console and type txData.securityProviderResponse, it should be undefined as in screenshot **Transaction Data 2**


## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

jpuri reviewed (changes requested):
We need to continue support opensea except in flask.

segun committed:
Deprecate open sea security alerts but only in flask

fix failing builds

Signed-off-by: Olusegun Akintayo <akintayo.segun@gmail.com>

Fix build issues

Fixed snaps e2e

return undefined not null for flask build

Signed-off-by: Olusegun Akintayo <akintayo.segun@gmail.com>

More code fencing.

Signed-off-by: Olusegun Akintayo <akintayo.segun@gmail.com>

Write a migration for blockaid
Fix actions and selector to remove check for old value.

Add to migration script, code fenced

More code fencing to fix errors in desktop and flask

Signed-off-by: Olusegun Akintayo <akintayo.segun@gmail.com>

Do not code fence, check for existence

Signed-off-by: Olusegun Akintayo <akintayo.segun@gmail.com>

link fixes

Fix flask build on code fencing

lint fixes

Delete 091.test.js

Fix build issues

Fix desktop build

Fixed snapshot

remove blockaid from build

Signed-off-by: Olusegun Akintayo <akintayo.segun@gmail.com>

Change migration to 94

Migration 93 from develop

Migration 095 is now for blockaid

add migration to index

Signed-off-by: Olusegun Akintayo <akintayo.segun@gmail.com>

Fix imports and code blocks

Fix lint
fix snapshot

Remove renderOpenSeaEnabledToggle from experimental settings.

Signed-off-by: Olusegun Akintayo <akintayo.segun@gmail.com>

Remove renderOpenSeaEnabledToggle from render

Signed-off-by: Olusegun Akintayo <akintayo.segun@gmail.com>

Lint fixes

jpuri commented:
@segun : code changes look good, I would suggest testing to ensure that `securityProviderResponse` is not added to any type of confirmation if opensea was previously enabled for a user in flask.

cryptotavares commented on app/scripts/metamask-controller.js:
```diff
@@ -4723,7 +4732,10 @@ export default class MetamaskController extends EventEmitter {
   };
 
-  async securityProviderRequest(requestData, methodName) {
+  // We can't code-fence the whole method as some controllers rely on the method. But we can return null
+  // if build is flask
```
I think that we can code-fence the entire `securityProviderRequest`. Currently it is used in:
* (TransactionController)[https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/transactions/index.js#L1544] -> it checks if `securityProviderRequest` is defined before calling it.
* SignatureController -> it is optional (see (here)[https://github.com/MetaMask/core/blob/main/packages/signature-controller/src/SignatureController.ts#L119C3-L119C26]). It passes the optional param to the AbstractMessageManager and there it checks if `securityProviderRequest` is defined before calling it (see (here)[https://github.com/MetaMask/core/blob/main/packages/message-manager/src/AbstractMessageManager.ts#L167]).
* MMIController - method does not seem to be used anywhere. Might be worth asking @zone-live about it
* createRPCMethodTrackingMiddleware -> (this)[https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/lib/createRPCMethodTrackingMiddleware.js#L198] is the only place that would require changes.. but I believe that we also need to apply codefencing there as well.

I don't think I have missed any usage.. but might be worth give it a second look.

segun replied:
The major issue is the MMIController one I guess. The whole MMI code is behind a code fence already. So can't add another code fence inline.

cryptotavares commented on app/scripts/controllers/metametrics.js:
```diff
@@ -727,6 +727,18 @@ export default class MetaMetricsController {
     const { traits, previousUserTraits } = this.store.getState();
+
+    const securityProviders = [];
+    if (metamaskState.securityAlertsEnabled) {
+      securityProviders.push('blockaid');
+    }
```
This could be code-fenced for blockaid

cryptotavares commented on ui/selectors/selectors.js:
```diff
@@ -1441,7 +1443,11 @@ export function getIsTransactionSecurityCheckEnabled(state) {
-  return state.metamask.securityAlertsEnabled;
+  // if the old settings is on, then return true else return false
+  return (
+    state.metamask.securityAlertsEnabled ||
+    state.metamask.transactionSecurityCheckEnabled
```
Why are we checking the old preference (`transactionSecurityCheckEnabled`)? Isn't that for the old security check `opensea + blockaid`?

segun replied:
There's a requirement to enable blockaid if the user has the old security check enabled.

cryptotavares replied:
This should be achieved with migrations and not by checking both properties... otherwise we will need to support the two options until a migration is done.

segun replied:
Yeah, doing the migration script now.

cryptotavares commented on ui/store/actions.ts:
```diff
@@ -4365,11 +4366,14 @@ export function setTransactionSecurityCheckEnabled(
   try {
     submitRequestToBackground('setSecurityAlertsEnabled', [val]);
+    // also update the old settings
+    submitRequestToBackground('setTransactionSecurityCheckEnabled', [val]);
```
Same as before. Why are we setting `setTransactionSecurityCheckEnabled`?

segun replied:
yes, because of the requirement to enable blockaid if old security check is enabled.

cryptotavares replied:
humm.. but this action is fired when a user explicit wants to enable the new blockaid security check, right? So why are we also setting the old property? That shouldn't be needed anymore right? 🤔

segun replied:
it could also be fired when user is turning off blockaid. But like Jyoti commented, doing a migration is better.

cryptotavares commented on app/scripts/metamask-controller.js:
```diff
@@ -4753,8 +4769,8 @@ export default class MetamaskController extends EventEmitter {
         const securityProviderResponse = await securityProviderCheck(
-          requestData,
-          methodName,
+          _requestData,
+          _methodName,
```
Now that code fencing is wrapping the entire method, this args do not need to be prefixed with `_`

cryptotavares commented on app/scripts/metamask-controller.js:
```diff
@@ -4742,7 +4757,8 @@ export default class MetamaskController extends EventEmitter {
   };
 
-  async securityProviderRequest(requestData, methodName) {
+  ///: BEGIN:ONLY_INCLUDE_IN(build-main,build-mmi,build-beta)
```
This looks better 👍 
But now you need to code fence the [createRPCMethodTrackingMiddleware](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/lib/createRPCMethodTrackingMiddleware.js#L116C25-L116C58) and it's usage of `securityProviderRequest`

segun replied:
code fencing[createRPCMethodTrackingMiddleware.js]( https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/lib/createRPCMethodTrackingMiddleware.js#L116C25-L116C58) is breaking jsdoc because code fencing is not working for jsdoc

I have added a check if the method securityProviderRequest exists before calling it.

jpuri commented:
@segun : is it possible to write a migration which will enable blockaid for all users which have opensea enabled currently.

cryptotavares commented on app/scripts/migrations/index.js:
```diff
@@ -185,6 +188,9 @@ const migrations = [
   m088,
   m089,
   m090,
+  ///: BEGIN:ONLY_INCLUDE_IN(blockaid)
+  m091,
+  ///: END:ONLY_INCLUDE_IN
```
as we have discussed, I am not sure that this is possible as it would could block or prevent subsequent migrations from running or if subsequent were executed before this one, then this one would never be executed. Example:
1. migration 91 only runs in Flask
2. a subsequent PR adds migration 92 and expects it to run everywhere (Flask, Beta, Main)
3. remove code fencing for blockaid and promote it to Main. At this time the versioned data is on 92. But on main it didn't run 91.. and it won't run it, because it will filter pending migrations to run with this [validation](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/lib/migrator/index.js#L77-L80). Aaand `versionedData.meta.version` is going to be 92.

The only way that I see it is to reference both `securityAlertsEnabled` and `transactionSecurityCheckEnabled` properties in Flask, but then we would do this migration once we were ready to promote to main.

@segun and @jpuri what do you think about it?

cryptotavares commented on ui/pages/settings/experimental-tab/experimental-tab.component.js:
```diff
@@ -6,11 +6,20 @@ import {
+  ///: END:ONLY_INCLUDE_IN
   TextColor,
   TextVariant,
+  ///: END:ONLY_INCLUDE_IN
```
This seems to be an extra `///: END:ONLY_INCLUDE_IN` directive and is causing the build to fail on the CI

cryptotavares commented on ui/pages/settings/experimental-tab/experimental-tab.component.js:
```diff
@@ -6,11 +6,20 @@ import {
 import {
-  FontWeight,
+  ///: BEGIN:ONLY_INCLUDE_IN(build-main,build-mmi,build-beta)
+  FONT_WEIGHT,
```
Should be `FontWeight` instead of `FONT_WEIGHT`... this is causing the lint to fail

cryptotavares commented on ui/pages/settings/experimental-tab/experimental-tab.component.js:
```diff
@@ -6,11 +6,20 @@ import {
 } from '../../../helpers/utils/settings-search';
 import { MetaMetricsEventCategory } from '../../../../shared/constants/metametrics';
+///: BEGIN:ONLY_INCLUDE_IN(build-main,build-mmi,build-beta)
+import Typography from '../../../components/ui/typography/typography';
```
`Typography` does not seem to be used anywhere.

cryptotavares commented on ui/pages/settings/experimental-tab/experimental-tab.component.js:
```diff
@@ -6,11 +6,20 @@ import {
   TextVariant,
+  ///: END:ONLY_INCLUDE_IN
+  ///: BEGIN:ONLY_INCLUDE_IN(build-main,build-mmi,build-beta)
+  TypographyVariant,
```
`TypographyVariant` does not seem to be used anywhere

segun committed:
Fix snapshot and build issues.

Signed-off-by: Olusegun Akintayo <akintayo.segun@gmail.com>

bschorchit commented:
We'll hold off on this migration
Additional context: [slack message](https://consensys.slack.com/archives/C04C0F6SSJH/p1692204350852359?thread_ts=1692052828.453279&cid=C04C0F6SSJH)
